File: AdjacentBit.js
var si = process.stdin;
var so = process.stdout;
var N = Number;
var _s = "";

function ch({ v, r, c }) {
  for (var x = 0; x < r; x++) {
    try {
      if (c > 1 && r > 1) {
        for (var y = 0; y < c; y++) {
          if (v[x][y] == v[x + 1][y] || v[x][y] == v[x][y + 1]) {
            return 0;
          }
        }
      } else {
        for (var t = 0; t < (c > 1 ? c : r); t++) {
          if (
            (c === 1 && v[x][0] === v[x + 1][0]) ||
            (r === 1 && v[0][t] === v[0][t + 1])
          ) {
            return 0;
          }
        }
      }
    } catch (e) {
      return 1;
    }
  }
  return 1;
}
function cr(r) {
  return r.split(" ");
}
function p(t) {
  so.write(t + "\n");
}

function m(input) {
  var l = input.split("\n");
  var t = N(l[0]),
    rs = [],
    pos = 1;
  for (var x = 0; x < t; x++) {
    var [r, c] = l[pos].split(" ");
    var np = pos + 1 + N(r);
    rs.push({
      r: N(r),
      c: N(c),
      v: l.slice(pos + 1, np).map(cr)
    });
    pos = np;
  }
  var s = rs.map(ch);
  s.forEach(p);
}
si.on("data", function(i) {
  _s += i;
});

si.on("end", function() {
  m(_s);
});

File: AdjacentBit.py
def i():return input().split(' ')
def R(a):return range(a)
def T(d,a,b):return d[a][b]==d[a+1][b]
def U(d,a,b):return d[a][b]==d[a][b+1]
def C(d,r,c):
    try:
        for a in R(r):
            if r>1 and c>1:
                for b in R(c): 
                    if T(d,a,b) or U(d,a,b):return 0
            else:
                for t in R(c if c>1 else r):
                    if (c==1 and T(d,a,0)) or (r==1 and U(d,0,t)):return 0
    except:return 1
for _ in R(int(input())):
    [r, c], d = map(int, i()), []
    for _ in R(r):d.append(i())
    print(C(d,r,c))
